File: nostos-core/src/test/java/sncf/oui/nostos/core/coordinates/LatitudeTest.java
package sncf.oui.nostos.core.coordinates;

import net.jqwik.api.*;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatThrownBy;

class LatitudeTest {

    @Property
    void out_of_range_degrees_should_throw_exception(@ForAll("invalidLatitudes") final double degrees) {
        assertThatThrownBy(() -> Latitude.of(degrees))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessage("latitude should be between -90° and 90°");
    }

    @Property
    void should_return_valid_latitude(@ForAll("validLatitudes") final double degrees) {
        final Latitude latitude = Latitude.of(degrees);

        assertThat(latitude.degrees()).isEqualTo(degrees);
    }

    @Provide
    Arbitrary<Double> invalidLatitudes() {
        final Arbitrary<Double> lowerBounds = Arbitraries.doubles().lessOrEqual(-90.).unique();
        final Arbitrary<Double> upperBounds = Arbitraries.doubles().greaterOrEqual(90.).unique();

        return Arbitraries.oneOf(lowerBounds, upperBounds);
    }

    @Provide
    Arbitrary<Double> validLatitudes() {
        return Arbitraries.doubles()
                .between(-90., 90.)
                .filter(latitude -> latitude != -90.)
                .filter(latitude -> latitude != 90.)
                .unique();
    }
}

File: nostos-core/src/test/kotlin/io/fries/nostos/core/coordinates/LatitudeTest.kt
package io.fries.nostos.core.coordinates

import net.jqwik.api.Arbitraries
import net.jqwik.api.ForAll
import net.jqwik.api.Property
import net.jqwik.api.Provide
import net.jqwik.api.constraints.DoubleRange
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatIllegalArgumentException

@Target(AnnotationTarget.VALUE_PARAMETER)
@Retention(AnnotationRetention.RUNTIME)
@DoubleRange(min = LatitudeTest.MIN_DEGREES, max = LatitudeTest.MAX_DEGREES)
internal annotation class LatitudeDegrees

internal class LatitudeTest {

    companion object {
        internal const val MIN_DEGREES = -90.0
        internal const val MAX_DEGREES = 90.0
    }

    @Property
    internal fun should_create_a_latitude_with_valid_degrees(@ForAll @LatitudeDegrees degrees: Double) {
        val latitude = Latitude(degrees)

        assertThat(latitude).isNotNull
        assertThat(latitude.degrees).isEqualTo(degrees)
    }

    @Property
    internal fun should_throw_when_the_degrees_are_out_of_range(@ForAll("out_of_range_degrees") degrees: Double) {
        assertThatIllegalArgumentException()
                .isThrownBy { Latitude(degrees) }
                .withMessage("Latitude degrees should be in the range [$MIN_DEGREES, $MAX_DEGREES]")
                .withNoCause()
    }

    @Provide
    fun out_of_range_degrees() = Arbitraries.oneOf(
            Arbitraries.doubles().lessOrEqual(MIN_DEGREES).filter { degrees -> degrees != MIN_DEGREES }.unique(),
            Arbitraries.doubles().greaterOrEqual(MAX_DEGREES).filter { degrees -> degrees != MAX_DEGREES }.unique()
    )
}